### ISSUE TITLE
db: better handling of background errors

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/pebble#270]
When a background error occurs during a flush or compaction, Pebble currently just logs the error and attempts to retry the operation indefinitely. Depending on the error which occurred, this isn't terribly helpful. If the error was due to lack of disk space, a retry loop can be helpful. But if the error is due to a logic bug in Pebble or some sort of corruption it would be better to place the the DB in a read-only mode. This is the strategy used by RocksDB.

Jira issue: PEBBLE-181

**Comment by petermattis** [comment]
@rohansuri has volunteered to take a look at this issue.

**Comment by rohansuri** [comment]
Overview of RocksDB's handling: https://github.com/facebook/rocksdb/wiki/Background-Error-Handling. I'll understand all the behaviours and put up the plan/any questions I have.

**Comment by rohansuri** [comment]
@petermattis Does the scope of this also cover to not panic on manifest write failures and continue serving reads? That will require the following changes: https://github.com/facebook/rocksdb/issues/6316, https://github.com/facebook/rocksdb/pull/5379.

**Comment by petermattis** [comment]
> @petermattis Does the scope of this also cover to not panic on manifest write failures and continue serving reads? That will require the following changes: facebook/rocksdb#6316, facebook/rocksdb#5379.

Yes, we'd want to be able to recover from out-of-disk errors during WAL writes, flushes/compactions, and MANIFEST writes.

**Comment by sumeerbhola** [comment]
We also need more visible error reporting so that we can quickly find errors and not have to scan a very verbose Pebble log. This came up when discussing a CockroachDB postmortem.
